plugin/trino-openlineage/src/main/java/io/trino/plugin/openlineage/OpenLineageListener.java

```diff
@@ -0,0 +1,312 @@
+ */
+package io.trino.plugin.openlineage;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+import io.openlineage.client.OpenLineage;
+import io.trino.spi.eventlistener.EventListener;
+import io.trino.spi.eventlistener.OutputColumnMetadata;
+import io.trino.spi.eventlistener.QueryCompletedEvent;
+import io.trino.spi.eventlistener.QueryContext;
+import io.trino.spi.eventlistener.QueryCreatedEvent;
+import io.trino.spi.eventlistener.QueryIOMetadata;
+import io.trino.spi.eventlistener.QueryMetadata;
+import io.trino.spi.eventlistener.QueryOutputMetadata;
+import io.trino.spi.eventlistener.QueryStatistics;
+
+import java.lang.reflect.Field;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class OpenLineageListener
+        implements EventListener
+{
+    private final OpenLineage ol = new OpenLineage(URI.create("https://github.com/trinodb/trino/plugin/trino-openlineage"));
+    private final OpenLineageEmitter client;
+    private final String trinoHost;
+    private final Integer trinoPort;
+    private final Boolean trinoMetadataFacetEnabled;
+    private final Boolean trinoQueryContextFacetEnabled;
+    private final Boolean queryStatisticsFacetEnabled;
+
+    @Inject
+    public OpenLineageListener(OpenLineageListenerConfig listenerConfig, OpenLineageClientConfig clientConfig)
+    {
+        this.client = new OpenLineageClient(clientConfig);
+
+        this.trinoHost = listenerConfig.getTrinoHost();
+        this.trinoPort = listenerConfig.getTrinoPort();
+
+        this.trinoMetadataFacetEnabled = listenerConfig.isMetadataFacetEnabled();
+        this.trinoQueryContextFacetEnabled = listenerConfig.isQueryContextFacetEnabled();
+        this.queryStatisticsFacetEnabled = listenerConfig.isQueryStatisticsFacetEnabled();
+    }
+
+    private UUID getQueryId(QueryMetadata queryMetadata)
+    {
+        return UUID.nameUUIDFromBytes(queryMetadata.getQueryId().getBytes(StandardCharsets.UTF_8));
+    }
+
+    @Override
+    public void queryCreated(QueryCreatedEvent queryCreatedEvent)
+    {
+        UUID runID = getQueryId(queryCreatedEvent.getMetadata());
+
+        try {
+            OpenLineage.RunEvent event = getStartEvent(runID, queryCreatedEvent);
+            client.emit(event, runID.toString());
+        }
+        catch (IllegalAccessException | JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void queryCompleted(QueryCompletedEvent queryCompletedEvent)
+    {
+        UUID runID = getQueryId(queryCompletedEvent.getMetadata());
+
+        try {
+            OpenLineage.RunEvent event = getCompletedEvent(runID, queryCompletedEvent);
```

Comment:
Shouldn't we just capture lineage for mutation queries i.e. DDLs, DMLs ?
Without [QueryType](https://github.com/trinodb/trino/blob/master/core/trino-spi/src/main/java/io/trino/spi/resourcegroups/QueryType.java#L21-L23) filtering, it is also going to emit lineage events for `SELECT` queries as well. Trino being a shared query engine for adhoc query analysis and dashboarding use cases, will normally see a huge volume of SELECT queries compared to DMLs or DDLs. So in the lineage backend system(e.g. Marquez) we will see a lot of lineage nodes for selects.


e.g. below set of queries will contribute 5 lineage events but from dataset lineage point of view we are interested to see the how `test_table` and `test_table2` got created from `nation`.
```
create table memory.default.test_table as select * from tpch.sf1.nation;
create table memory.default.test_table2 as select * from tpch.sf1.nation;
select * from from tpch.sf1.nation;
select * from from tpch.sf1.nation;
select * from from tpch.sf1.nation;
```


Dataset owners would rather prefer to see which ddl/dml queries are mutating their datasets and would be interested to see respective upstream and downstream lineage events, as compared no-transformation vanilla select queries.

`queryCompletedEvent.getContext().getQueryType()` will get you the `QueryType` here.

We can also think about giving this option back to user, so they can dictate if they are interested in seeing lineage for all query types or just a subset. That configuration can go to `OpenLineageListenerConfig`

Reply:
very good point!

Reply:
addressed here https://github.com/mgorsk1/trino/pull/1

Reply:
I think this is something that could be configurable by plugin user - not sure how Trino plugins do it - but getting events for `SELECT` queries is a feature that some people would like.

Reply:
yep, I made it into a config option with sensible defaults